internal/export/ingest_pipelines.go

```diff
@@ -0,0 +1,131 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package export
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/elastic/elastic-package/internal/elasticsearch"
+	"github.com/elastic/elastic-package/internal/elasticsearch/ingest"
+)
+
+type PipelineWriteLocationType string
+
+const (
+	PipelineWriteLocationTypeDataStream PipelineWriteLocationType = "data_stream"
+	PipelineWriteLocationTypeRoot       PipelineWriteLocationType = "root"
+)
+
+// Represents a target write location for exporting an ingest pipeline
+type PipelineWriteLocation struct {
+	Type       PipelineWriteLocationType
+	Name       string
+	ParentPath string
+}
+
+func (p PipelineWriteLocation) WritePath() string {
+	return filepath.Join(p.ParentPath, "elasticsearch", "ingest_pipeline")
+}
+
+type PipelineWriteAssignments map[string]PipelineWriteLocation
+
+func IngestPipelines(ctx context.Context, api *elasticsearch.API, writeAssignments PipelineWriteAssignments) error {
+	var pipelineIDs []string
+
+	for pipelineID := range writeAssignments {
+		pipelineIDs = append(pipelineIDs, pipelineID)
+	}
+
+	pipelines, err := ingest.GetRemotePipelinesWithNested(ctx, api, pipelineIDs...)
+
+	if err != nil {
+		return fmt.Errorf("exporting ingest pipelines using Elasticsearch failed: %w", err)
+	}
+
+	pipelineLookup := make(map[string]ingest.RemotePipeline)
+	for _, pipeline := range pipelines {
+		pipelineLookup[pipeline.Name()] = pipeline
+	}
+
+	err = writePipelinesToFiles(writeAssignments, pipelineLookup)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func writePipelinesToFiles(writeAssignments PipelineWriteAssignments, pipelineLookup map[string]ingest.RemotePipeline) error {
+	if len(writeAssignments) == 0 {
+		return nil
+	}
+
+	for name, writeLocation := range writeAssignments {
+		pipeline, ok := pipelineLookup[name]
+		if !ok {
+			continue
+		}
+		err := writePipelineToFile(pipeline, writeLocation)
+		if err != nil {
+			return err
+		}
+
+		depPipelineWriteAssignments := createWriteAssignments(writeLocation, pipeline.GetProcessorPipelineNames())
+		err = writePipelinesToFiles(depPipelineWriteAssignments, pipelineLookup)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func writePipelineToFile(pipeline ingest.RemotePipeline, writeLocation PipelineWriteLocation) error {
+	var jsonPipeline map[string]any
+	err := json.Unmarshal(pipeline.JSON(), &jsonPipeline)
+	if err != nil {
+		return fmt.Errorf("unmarshalling ingest pipeline failed (ID: %s): %w", pipeline.Name(), err)
+	}
+
+	yamlBytes, err := yaml.Marshal(jsonPipeline)
+	if err != nil {
+		return fmt.Errorf("marshalling ingest pipeline json to yaml failed (ID: %s): %w", pipeline.Name(), err)
+	}
+
+	// requirement: https://github.com/elastic/package-spec/pull/54
+	documentStartDashes := []byte("---\n")
+	documentBytes := append(documentStartDashes, yamlBytes...)
```

Comment:
Nit. To avoid appending, you can use a bytes buffer. Something like this:

```suggestion
	var documentBytes bytes.Buffer
	// requirement: https://github.com/elastic/package-spec/pull/54
	documentBytes.WriteString("---\n")
	err := yaml.NewEncoder(&documentBytes).Encode(jsonPipeline)
	if err != nil {
		return fmt.Errorf("marshalling ingest pipeline json to yaml failed (ID: %s): %w", pipeline.Name(), err)
	}
```